extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,618 @@
+    attributes['sentinel'] = this.sentinel_;
+    return attributes;
+  }
+
+  /**
+   * Returns the initialGeometry to assign to the name of the safeframe
+   * for rendering. This needs to be done differently than all the other
+   * geometry updates, because we don't actually have access to the
+   * rendered safeframe yet. Note that we are using getPageLayoutBox,
+   * which is not guaranteed to be perfectly accurate as it is from
+   * the last measure of the element. This is fine for our use case
+   * here, as even if the position is slightly off, we'll send the right
+   * size.
+   * @return {string}
+   */
+  getInitialGeometry() {
+    const ampAdBox = this.baseInstance_.getPageLayoutBox();
+    const heightOffset = (ampAdBox.height - this.creativeSize_.height) / 2;
+    const widthOffset = (ampAdBox.width - this.creativeSize_.width) / 2;
+    const iframeBox = {
+      top: ampAdBox.top + heightOffset,
+      bottom: ampAdBox.bottom - heightOffset,
+      left: ampAdBox.left + widthOffset,
+      right: ampAdBox.right - widthOffset,
+      height: this.creativeSize_.height,
+      width: this.creativeSize_.width,
+    };
+    return this.formatGeom_(iframeBox);
+  }
+
+  /**
+   * Registers this instance as the host API for the current sentinel.
+   * If the global safeframe listener has not yet been created, it creates
+   * that as well.
+   */
+  registerSafeframeHost() {
+    dev().assert(this.sentinel_);
+    safeframeHosts[this.sentinel_] = safeframeHosts[this.sentinel_] || this;
+    if (!safeframeListenerCreated_) {
+      safeframeListenerCreated_ = true;
+      this.win_.addEventListener('message', safeframeListener, false);
+    }
+  }
+
+  /**
+   * Sends initial connection message to the safeframe to finish initialization.
+   * Also initializes the sending of geometry update messages to the frame.
+   * @param {string} channel
+   */
+  connectMessagingChannel(channel) {
+    // Set the iframe here, because when class is first created the iframe
+    // element does not yet exist on this.baseInstance_. The first time
+    // we receive a message we know that it now exists.
+    dev().assert(this.baseInstance_.iframe);
+    this.iframe_ = this.baseInstance_.iframe;
+    this.channel = channel;
+    this.setupGeom_();
+    this.sendMessage_({
+      'message': 'connect',
+      'c': this.channel,
+    }, '');
+  }
+
+  /**
+   * Creates IntersectionObserver instance for this SafeframeAPI instance.
+   * We utilize the existing IntersectionObserver class, by passing in this
+   * class for IO to use instead of SubscriptionApi for sending its update
+   * messages. The method 'send' below is triggered by IO every time that
+   * an update occurs.
+   * @private
+   */
+  setupGeom_() {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    const throttledUpdate = throttle(
+        this.win_, this.updateGeometry_.bind(this), 1000);
+    this.viewport_.onScroll(throttledUpdate);
```

Comment:
This may need to be cleanup on unlayoutCallback as well

Reply:
done